cranelift/codegen/src/machinst/mod.rs

```diff
@@ -272,13 +273,18 @@ pub trait MachInstEmit: MachInst {
 /// emitting a function body.
 pub trait MachInstEmitState<I: VCodeInst>: Default + Clone + Debug {
     /// Create a new emission state given the ABI object.
-    fn new(abi: &Callee<I::ABIMachineSpec>) -> Self;
+    fn new(abi: &Callee<I::ABIMachineSpec>, ctrl_plane: ControlPlane) -> Self;
     /// Update the emission state before emitting an instruction that is a
     /// safepoint.
     fn pre_safepoint(&mut self, _stack_map: StackMap) {}
     /// Update the emission state to indicate instructions are associated with a
     /// particular RelSourceLoc.
     fn pre_sourceloc(&mut self, _srcloc: RelSourceLoc) {}
+    /// The emission state holds temporary ownership of a control plane. A
+    /// mutable reference to it may be used to:
+    /// - move out of it, e.g. using [std::mem::swap]
+    /// - access that control plane temporarily
+    fn get_ctrl_plane(&mut self) -> &mut ControlPlane;
```

Comment:
A few things:
- Usually a mut-accessor will be named like `fn ctrl_plane_mut(&mut self) -> ...`
- Let's have a different one too, `fn take_ctrl_plane(self)`, that consumes the emit-state and gives us back the control-plane state

Reply:
> `fn take_ctrl_plane(self)`

I think that actually caught a mistake. I was taking the control plane out of the emission state inside a loop, where later loop iterations would use the state with a now-empty control plane.